File: internal/runtime/stats.go
package runtime

import (
	"sync"
	"sync/atomic"
	"time"
)

// OpStats holds per-operator cumulative execution statistics.
// All fields are updated atomically and are safe for concurrent access.
type OpStats struct {
	ExecCount      int64 // number of successful executions
	SkipCount      int64 // number of times the operator was skipped
	ErrorCount     int64 // number of errors (fatal)
	TotalDurationNs int64 // cumulative execution time in nanoseconds
	MaxDurationNs   int64 // maximum single-execution time in nanoseconds
}

// Stats accumulates per-operator execution statistics across requests.
// Thread-safe for concurrent updates from multiple Execute calls.
type Stats struct {
	mu    sync.RWMutex
	ops   map[string]*OpStats
}

// NewStats creates a new Stats accumulator.
func NewStats() *Stats {
	return &Stats{
		ops: make(map[string]*OpStats),
	}
}

// getOrCreate returns the OpStats for a given operator, creating it if needed.
func (s *Stats) getOrCreate(name string) *OpStats {
	s.mu.RLock()
	st, ok := s.ops[name]
	s.mu.RUnlock()
	if ok {
		return st
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	// Double-check after acquiring write lock
	if st, ok = s.ops[name]; ok {
		return st
	}
	st = &OpStats{}
	s.ops[name] = st
	return st
}

// RecordExec records a successful operator execution.
func (s *Stats) RecordExec(name string, duration time.Duration) {
	st := s.getOrCreate(name)
	atomic.AddInt64(&st.ExecCount, 1)
	ns := duration.Nanoseconds()
	atomic.AddInt64(&st.TotalDurationNs, ns)

	// CAS loop for max
	for {
		cur := atomic.LoadInt64(&st.MaxDurationNs)
		if ns <= cur {
			break
		}
		if atomic.CompareAndSwapInt64(&st.MaxDurationNs, cur, ns) {
			break
		}
	}
}

// RecordSkip records a skipped operator execution.
func (s *Stats) RecordSkip(name string) {
	st := s.getOrCreate(name)
	atomic.AddInt64(&st.SkipCount, 1)
}

// RecordError records a failed operator execution.
func (s *Stats) RecordError(name string, duration time.Duration) {
	st := s.getOrCreate(name)
	atomic.AddInt64(&st.ErrorCount, 1)
	ns := duration.Nanoseconds()
	atomic.AddInt64(&st.TotalDurationNs, ns)

	for {
		cur := atomic.LoadInt64(&st.MaxDurationNs)
		if ns <= cur {
			break
		}
		if atomic.CompareAndSwapInt64(&st.MaxDurationNs, cur, ns) {
			break
		}
	}
}

// Snapshot returns a point-in-time copy of all operator statistics.
type OpStatsSnapshot struct {
	ExecCount      int64   `json:"exec_count"`
	SkipCount      int64   `json:"skip_count"`
	ErrorCount     int64   `json:"error_count"`
	TotalDurationNs int64  `json:"total_duration_ns"`
	MaxDurationNs   int64  `json:"max_duration_ns"`
	AvgDurationNs   int64  `json:"avg_duration_ns"`
}

// Snapshot returns a point-in-time copy of all operator statistics.
func (s *Stats) Snapshot() map[string]OpStatsSnapshot {
	s.mu.RLock()
	defer s.mu.RUnlock()

	result := make(map[string]OpStatsSnapshot, len(s.ops))
	for name, st := range s.ops {
		exec := atomic.LoadInt64(&st.ExecCount)
		total := atomic.LoadInt64(&st.TotalDurationNs)
		var avg int64
		if exec > 0 {
			avg = total / exec
		}
		result[name] = OpStatsSnapshot{
			ExecCount:       exec,
			SkipCount:       atomic.LoadInt64(&st.SkipCount),
			ErrorCount:      atomic.LoadInt64(&st.ErrorCount),
			TotalDurationNs: total,
			MaxDurationNs:   atomic.LoadInt64(&st.MaxDurationNs),
			AvgDurationNs:   avg,
		}
	}
	return result
}

File: pine-java/src/main/java/page/liam/pine/Stats.java
package page.liam.pine;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

public class Stats {
    private final ConcurrentHashMap<String, OpStats> ops = new ConcurrentHashMap<>();
    private final AtomicLong runCount = new AtomicLong();
    private final AtomicLong peakConcurrency = new AtomicLong();

    public void preInitOperators(List<String> names) {
        for (String name : names) {
            ops.putIfAbsent(name, new OpStats());
        }
    }

    public void recordExec(String name, long durationNs) {
        getOrCreate(name).recordExec(durationNs);
    }

    public void recordSkip(String name) {
        getOrCreate(name).skipCount.incrementAndGet();
    }

    public void recordError(String name, long durationNs) {
        getOrCreate(name).recordError(durationNs);
    }

    public void recordRun() {
        runCount.incrementAndGet();
    }

    public void recordConcurrency(long n) {
        long prev;
        do {
            prev = peakConcurrency.get();
            if (n <= prev) return;
        } while (!peakConcurrency.compareAndSet(prev, n));
    }

    public Map<String, Map<String, Object>> snapshot() {
        Map<String, Map<String, Object>> result = new TreeMap<>();
        for (Map.Entry<String, OpStats> e : ops.entrySet()) {
            result.put(e.getKey(), e.getValue().snapshot());
        }
        return result;
    }

    public Map<String, Object> schedulerSnapshot() {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("run_count", runCount.get());
        m.put("peak_concurrency", peakConcurrency.get());
        return m;
    }

    private OpStats getOrCreate(String name) {
        return ops.computeIfAbsent(name, k -> new OpStats());
    }

    public static class OpStats {
        final AtomicLong execCount = new AtomicLong();
        final AtomicLong skipCount = new AtomicLong();
        final AtomicLong errorCount = new AtomicLong();
        final AtomicLong totalDurationNs = new AtomicLong();
        final AtomicLong maxDurationNs = new AtomicLong();

        void recordExec(long durationNs) {
            execCount.incrementAndGet();
            totalDurationNs.addAndGet(durationNs);
            long prev;
            do {
                prev = maxDurationNs.get();
                if (durationNs <= prev) return;
            } while (!maxDurationNs.compareAndSet(prev, durationNs));
        }

        void recordError(long durationNs) {
            errorCount.incrementAndGet();
            totalDurationNs.addAndGet(durationNs);
            long prev;
            do {
                prev = maxDurationNs.get();
                if (durationNs <= prev) return;
            } while (!maxDurationNs.compareAndSet(prev, durationNs));
        }

        Map<String, Object> snapshot() {
            long exec = execCount.get();
            long total = totalDurationNs.get();
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("exec_count", exec);
            m.put("skip_count", skipCount.get());
            m.put("error_count", errorCount.get());
            m.put("total_duration_ns", total);
            m.put("max_duration_ns", maxDurationNs.get());
            m.put("avg_duration_ns", exec > 0 ? total / exec : 0L);
            return m;
        }
    }
}
